owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt

```diff
@@ -190,18 +193,50 @@ class DocumentsStorageProvider : DocumentsProvider() {
         val file = currentStorageManager?.getFileById(docId) ?: throw FileNotFoundException("File $docId not found")
         Log_OC.d(TAG, "Trying to rename ${file.fileName} to $displayName")
 
-        val renameFileOperation = RenameFileOperation(file.remotePath, displayName)
-        val result = renameFileOperation.execute(currentStorageManager, context)
+        RenameFileOperation(file.remotePath, displayName).apply {
+            execute(currentStorageManager, context).also { checkOperationResult(it, file) }
+        }
+
+        return null
+    }
+
+    override fun deleteDocument(documentId: String) {
+        val docId = documentId.toLong()
+
+        updateCurrentStorageManagerIfNeeded(docId)
+
+        val file = currentStorageManager?.getFileById(docId) ?: throw FileNotFoundException("File $docId not found")
+        Log_OC.d(TAG, "Trying to delete ${file.fileName} with id ${file.fileId}")
+
+        RemoveFileOperation(file.remotePath, false).apply {
+            execute(currentStorageManager, context).also { checkOperationResult(it, file) }
+        }
+    }
+
+    @TargetApi(24)
+    override fun removeDocument(documentId: String, parentDocumentId: String) {
+
+        //Documents can only have one parent
+        deleteDocument(documentId)
+
+    }
+
+    private fun checkOperationResult(result: RemoteOperationResult<Any>, file: OCFile) {
 
         if (!result.isSuccess) {
+
             context?.contentResolver?.notifyChange(toNotifyUri(toUri(file.parentId.toString())), null)
-            throw java.lang.UnsupportedOperationException("Rename failed")
+            throw FileNotFoundException("Remote Operation failed due to ${result.exception.message}")
+
         } else {
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                revokeDocumentPermission(file.fileId.toString())
+            }
             syncRequired = false
             context?.contentResolver?.notifyChange(toNotifyUri(toUri(file.parentId.toString())), null)
```

Comment:
This line appears several times in the class, you might create a new method called `notifyChangeInDocument(documentId)` or something like that

Reply:
Ok